sonic-net/sonic-buildimage
# [dhcp_server] Add dhcprelayd for dhcp_server feature

#### Why I did it
Add support in dhcp_relay container for dhcp_server_ipv4 feature. HLD: https://github.com/sonic-net/SONiC/pull/1282

##### Work item tracking
- Microsoft ADO **(number only)**: 25506540

#### How I did it
* Inside dhcp_relay container:
1. Add `install psutil` and `install sonic_dhcp_server` python wheels in Dockerfile.j2
2. For dhcp_relay cli, add restrictions for disabling add/del/show ipv4 dhcp_relay while dhcp_server is enabled.
3. Add dhcprelayd process, to start/restart `dhcpmon` (dhcpv4 relay counter) and `dchrelay` processes. And monitor config_db change.
4. Modify template to generate supervisord.conf file. If dhcp_server feature is disabled, generate configuration file as before. If dhcp_server feature is enabled, make supervisor not run `dhcpmon` (dhcpv4 relay counter) and `dchrelay` process, instead run dhcprelayd process. 
5. Add DhcpRelaydDbMonitor to monitor db changes
* Inside dhcp_server container:
1. Add support for dhcp_lease to remove expired lease found in lease file.
2. After init, add dhcp_server_ip (eth0 inside dhcp_server container) to state_db
* Inside sonic-config-engine wheel
1. Add/Modify UTs for configuration j2 for dhcp_relay changes.
1.1 For dhcp_server disabled, remove repeated blank lines in expected output file.
1.2 For dhcp_server enabled, add new UT.
* For sonic-dhcp-server python wheel
1. Add dhcp_db_monitor.py which is used to monitor db change.
2. Refactor folder structure inside src/sonic-dhcp-server/dhcp_server

#### How to verify it
* UTs in sonic-config-engine all passed, to make sure that supervisor configuration file for dhcp_relay container is generated correct.
* UTs in sonic-dhcp-server all passed
```
=========================================================== test session starts ============================================================
platform linux -- Python 3.9.2, pytest-6.0.2, py-1.10.0, pluggy-0.13.0
rootdir: /sonic/src/sonic-dhcp-server, configfile: setup.cfg, testpaths: tests
plugins: pyfakefs-5.2.4, cov-2.10.1
collected 140 items                                                                                                                        

tests/test_dhcp_cfggen.py ..........                                                                                                 [  7%]
tests/test_dhcp_db_monitor.py .......................                                                                                [ 23%]
tests/test_dhcp_lease.py .....                                                                                                       [ 27%]
tests/test_dhcprelayd.py ........................................................................................                    [ 90%]
tests/test_dhcpservd.py ......                                                                                                       [ 94%]
tests/test_utils.py ........                                                                                                         [100%]Clearing the cache


----------- coverage: platform linux, python 3.9.2-final-0 -----------
Name                                    Stmts   Miss Branch BrPart     Cover   Missing
--------------------------------------------------------------------------------------
dhcp_server/dhcprelayd/dhcprelayd.py      120      8     48      0    91.67%   73-82
dhcp_server/dhcpservd/dhcp_cfggen.py      169     16     72      1    92.95%   48-65, 148-152, 238->240, 240
dhcp_server/dhcpservd/dhcp_lease.py        95      3     30      4    94.40%   57->59, 59, 60->61, 61, 74->76, 112->134, 153
dhcp_server/common/utils.py                40      3     14      0    94.44%   35, 55-56
dhcp_server/common/dhcp_db_monitor.py      81      1     30      1    98.20%   69->70, 70
dhcp_server/dhcpservd/dhcpservd.py         49      1     10      0    98.31%   73
--------------------------------------------------------------------------------------
TOTAL                                     554     32    204      6    94.20%

1 file skipped due to complete coverage.

Required test coverage of 80.0% reached. Total coverage: 94.20%

=========================================================== slowest 5 durations ============================================================
0.07s call     tests/test_dhcp_lease.py::test_update_kea_lease
0.02s call     tests/test_dhcprelayd.py::test_start_dhcpmon_process[zombie-2-new_dhcp_interfaces_list0]
0.01s call     tests/test_dhcp_cfggen.py::test_render_config[True]
0.01s call     tests/test_dhcpservd.py::test_dump_dhcp4_config
0.01s call     tests/test_dhcp_cfggen.py::test_render_config[False]
```


#### Which release branch to backport (provide reason below if selected)




#### Tested branch (Please provide the tested image version)




#### Description for the changelog




#### Link to config_db schema for YANG module changes


#### A picture of a cute animal (not mandatory but encouraged)
